course/src/ant.cc

```diff
@@ -0,0 +1,53 @@
+#include "insect.h"
+#include "ant.h"
+
+namespace ekumen {
+namespace simulation {
+
+Ant::Ant(std::shared_ptr<Cell>& cell) {
+    this->cell = cell;
+    cell->Occupy(std::move(GetThisPtr()));
+    metrics.SetInsectType(Ant::GetInsectType());
+}
+
+InsectType Ant::GetInsectType() {
+    return InsectType::Ant;
+}
+
+void Ant::RunRound() {
+    if (IsDead()) {
+        metrics.SetDead();
+    }
+    else {
+        has_bred_current_round = false;
+        Breed();
+        Move();
+        // If the ant wasn't able to breed before moving, it should
+        // check if it can after moving.
+        // This boosts the overall survival rate of the ants.
+        Breed();
+        UpdateBreedingState();
+    }
+    if(round_finished_callback != nullptr) {
+        round_finished_callback(metrics);
+    }
+}
+
+uint32_t Ant::GetRequiredRoundsToBreed() {
+    return kRequiredRoundsUntilBreed;
+}
+
+std::shared_ptr<Insect> Ant::GetNewborn(std::shared_ptr<Cell>& cell) {
+    std::shared_ptr<Ant> newborn(new Ant(cell));
+    return std::move(newborn);
+}
+
+std::shared_ptr<Cell> Ant::WhereCanIEat() {
+    return nullptr;
+}
+
+std::shared_ptr<Insect> Ant::GetThisPtr() {
+    return std::make_shared<Ant>(*this);
```

Comment:
Why so?

Reply:
Because you are creating a new shared pointer from a raw pointer. You should use the method `shared_from_this()` provided by the `shared_from_this` class. See [here](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this/shared_from_this).

Reply:
After talking off-thread, I'll address the fix for this issue in an upcoming PR.